File: contract/src/infrastructure/messaging/rabbitmq-connectie.ts
import amqp, { type ChannelModel, type Channel } from 'amqplib';

const EXCHANGE = 'rws.events';

export class RabbitMqConnectie {
  private constructor(
    private readonly verbinding: ChannelModel,
    private readonly ch: Channel,
  ) {}

  static async verbind(url: string): Promise<RabbitMqConnectie> {
    const verbinding = await amqp.connect(url);
    const ch = await verbinding.createChannel();
    await ch.assertExchange(EXCHANGE, 'topic', { durable: true });
    return new RabbitMqConnectie(verbinding, ch);
  }

  get kanaal(): Channel {
    return this.ch;
  }

  isVerbonden(): boolean {
    return this.ch !== undefined;
  }

  async sluit(): Promise<void> {
    await this.ch.close();
    await this.verbinding.close();
  }
}

export const RWS_EXCHANGE = EXCHANGE;

File: monitoring/src/Monitoring.Infrastructure/Messaging/RabbitMqConnectie.cs
using RabbitMQ.Client;

namespace Monitoring.Infrastructure.Messaging;

public sealed class RabbitMqConnectie : IAsyncDisposable
{
    public const string Exchange = "rws.events";

    private readonly IConnection _connectie;
    public IChannel Kanaal { get; }

    private RabbitMqConnectie(IConnection connectie, IChannel kanaal)
    {
        _connectie = connectie;
        Kanaal = kanaal;
    }

    public static async Task<RabbitMqConnectie> VerbindAsync(string url)
    {
        var factory = new ConnectionFactory { Uri = new Uri(url) };
        var connectie = await factory.CreateConnectionAsync();
        var kanaal = await connectie.CreateChannelAsync();
        await kanaal.ExchangeDeclareAsync(Exchange, ExchangeType.Topic, durable: true);
        return new RabbitMqConnectie(connectie, kanaal);
    }

    public bool IsVerbonden => _connectie.IsOpen && Kanaal.IsOpen;

    public Task<IChannel> MaakKanaalAsync() => _connectie.CreateChannelAsync();

    public async ValueTask DisposeAsync()
    {
        try { await Kanaal.CloseAsync(); } catch { /* al gesloten */ }
        try { await _connectie.CloseAsync(); } catch { /* al gesloten */ }
        _connectie.Dispose();
    }
}
